frametrim/ft_d3d11.hpp

```diff
@@ -0,0 +1,692 @@
+public:
+    D3D11Texture1D(ImplPtr impl, void *id);
+
+    virtual unsigned getMappedSize(unsigned width_stride, unsigned depth_stride);
+};
+
+class D3D11Texture2D: public D3D11Resource {
+public:
+    D3D11Texture2D(ImplPtr impl, void *id, unsigned height, unsigned bind_flags);
+
+    virtual unsigned getMappedSize(unsigned width_stride, unsigned depth_stride);
+    virtual bool hasBindFlag(unsigned flag) { return m_bind_flags & flag; };
+
+private:
+    unsigned m_height;
+    unsigned m_bind_flags;
+};
+
+class D3D11Texture3D: public D3D11Resource {
+public:
+    D3D11Texture3D(ImplPtr impl, void *id, unsigned depth);
+
+    virtual unsigned getMappedSize(unsigned width_stride, unsigned depth_stride);
+
+private:
+    unsigned m_depth;
+};
+
+class D3D11View: public D3D11DeviceChild {
+public:
+    using Pointer = std::shared_ptr<D3D11View>;
+
+    D3D11View(ImplPtr impl, void *id, D3D11Resource::Pointer res);
+
+    void addUpdateCall(const trace::Call& call);
+    void addUpdateDependency(Object::Pointer dep);
+    void clear(const trace::Call& call);
+    bool hasBindFlag(unsigned flag) { return m_resource->hasBindFlag(flag); };
+
+private:
+    D3D11Resource::Pointer m_resource;
+};
+
+class D3D11Operation: public D3D11DeviceChild {
+public:
+    using Pointer = std::shared_ptr<D3D11Operation>;
+
+    D3D11Operation(ImplPtr impl, void *id);
+
+    uint32_t frameNo() const { return m_frame_no; };
+    virtual void execute(uint32_t frame_no) = 0;
+
+protected:
+    uint32_t m_frame_no;
+};
+
+class D3D11ResourceUpdate: public D3D11Operation {
+public:
+    using Pointer = std::shared_ptr<D3D11Operation>;
+
+    D3D11ResourceUpdate(ImplPtr impl, void *id, D3D11Resource::Pointer res, D3D11Mapping& mapping);
+    virtual void execute(uint32_t frame_no);
+
+private:
+    D3D11Resource::Pointer m_res;
+    D3D11Mapping m_mapping;
+};
+
+using SubresourceId = std::tuple<void *, void *, unsigned>;
+using MappingMap = std::map<SubresourceId, D3D11Mapping>;
+typedef bool(*FilterFn)(unsigned, unsigned, D3D11DeviceChild::Pointer obj);
+
+class D3D11Command: public D3D11Operation {
+public:
+    using Pointer = std::shared_ptr<D3D11Command>;
+
+    D3D11Command(ImplPtr impl, void *id, bool deferred);
+
+    virtual void link(States& states, Bindings& bindings) = 0;
+    virtual void execute(uint32_t frame_no);
+
+protected:
+    void addStates(States& states);
+    void addBoundAsDependency(Bindings& bindings, enum ePerContextBinding pcb, bool compute);
+    void addToBound(Bindings& bindings, enum ePerContextBinding pcb);
+    void addToBoundView(Bindings& bindings, enum ePerContextBinding pcb, bool compute, FilterFn filter);
+
+private:
+    bool m_deferred;
+    std::vector<D3D11View::Pointer> m_delayed_targets;
+};
+
+class D3D11Draw: public D3D11Command {
+public:
+    using Pointer = std::shared_ptr<D3D11Draw>;
+
+    D3D11Draw(ImplPtr impl, void *id, bool deferred);
+    virtual void link(States& states, Bindings& bindings);
```

Comment:
Style: better use `void link(States& states, Bindings& bindings) override` with that the compiler will complain when the base class doesn't have this virtual method.